### PR TITLE
Add CI

### PR DESCRIPTION
**Pull request by martijnbastiaan**
Shamelessly stolen from https://github.com/DigitalBrains1/prettyprinter-interp

### CODE REVIEW

File: .ci/cabal.project.local-lower [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package docopt
+  ghc-options: -Werror
+
+prefer-oldest: True
```
Comment (DigitalBrains1) [inline .ci/cabal.project.local-lower:4]: Needs:
```
constraints:
  aeson >= 1.4.4.0,
  contravariant >= 1.5.2
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    name: Cabal / GHC ${{ matrix.ghc }} ${{ matrix.project-variant }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        # TODO: "9.2.8", "9.4.7", "9.6.3"
+        ghc: ["8.0.2", "8.2.2", "8.4.4", "8.6.5", "8.8.4", "8.10.7", "9.0.2"]
+        project-variant: [""]
+        include:
+          - ghc: 8.6.5
+            project-variant: -lower
```
Comment (DigitalBrains1) [inline .github/workflows/ci.yml:24]: My suggestion is to do this for GHC 8.0.2 instead, test the oldest of the oldest.

Comment (martijnbastiaan) [reply .github/workflows/ci.yml:24]: I _just_ noticed, one sec

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            'dist-newstyle/cache/plan.json') }}
+          restore-keys: ${{ env.key }}-
+
+      - name: Install dependencies
+        run: cabal v2-build all --enable-tests --only-dependencies
+
+      # Cache dependencies already at this point, so that we do not have to
+      # rebuild them should the subsequent steps fail
+      - name: Save cached dependencies
+        uses: actions/cache/save@v3
```
Comment (DigitalBrains1) [inline .github/workflows/ci.yml:80]:
```suggestion
        uses: actions/cache/save@v4
```
and the other locations where `actions/cache` is used; there are deprecation warnings now.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        uses: actions/cache/save@v4
+        # Trying to save over an existing cache gives distracting
+        # "Warning: Cache save failed." since they are immutable
+        if: steps.cache.outputs.cache-hit != 'true'
+        with:
+          path: ${{ steps.setup-haskell.outputs.cabal-store }}
+          key: ${{ steps.cache.outputs.cache-primary-key }}
+
+      - name: Build
+        run: cabal v2-build all
```
Comment (DigitalBrains1) [inline .github/workflows/ci.yaml:161]:
```suggestion
        run: cabal v2-build all --enable-tests
```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+      - name: Test
+        run: cabal v2-test --test-show-details=direct
```
Comment (DigitalBrains1) [inline .github/workflows/ci.yaml:164]:
```suggestion
        run: cabal v2-test --enable-tests --test-show-details=direct
```
Somehow only `-lower` was affected, but this does fix it.

File: .ci/stack-9.6.yaml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+resolver: nightly-2023-10-09
```
Comment (DigitalBrains1) [inline .ci/stack-9.6.yaml:1]: Let's pick LTS-22.11 and add a nightly for GHC 9.8.

**Pull request merged**